### PR TITLE
fix(expr): fix common_super_type() for decimal

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Closes #issue

### CODE REVIEW

**Review by youngsofun — APPROVED** [review]

File: src/query/expression/src/type_check.rs [hunk 1]
```diff
@@ -559,24 +558,24 @@ pub fn common_super_type(
...
-
+        (DataType::Number(num_ty), decimal_ty @ DataType::Decimal(_))
+        | (decimal_ty @ DataType::Decimal(_), DataType::Number(num_ty))
+            if !num_ty.is_float() =>
+        {
+            Some(decimal_ty)
+        }
+        (DataType::Number(num_ty), DataType::Decimal(_))
+        | (DataType::Decimal(_), DataType::Number(num_ty))
+            if num_ty.is_float() =>
```
Comment (sundy-li) [inline src/query/expression/src/type_check.rs:575]: if `num_ty` is not float, should be decimal of `max_precision`.

Comment (youngsofun) [reply src/query/expression/src/type_check.rs:575]: `range > range(num_ty)` is enough?

Comment (andylokandy) [reply src/query/expression/src/type_check.rs:575]: > if num_ty is not float, should be decimal of max_precision.

if so, we should carefully compare the decimal in `can_auto_cast_to`: https://github.com/datafuselabs/databend/blob/45f3ca1bba14f1d192f9ab0094a9f63e9acfe792/src/query/expression/src/type_check.rs#L521

Comment (sundy-li) [reply src/query/expression/src/type_check.rs:575]: But it's always true, because `10**38` > `2^64 - 1`

Comment (andylokandy) [reply src/query/expression/src/type_check.rs:575]: I mean, for `(DataType::Number(n), DataType::Decimal(DecimalSize{precision: 1, scale: 1}))` , `can_auto_cast_to` should return false

Comment (sundy-li) [reply src/query/expression/src/type_check.rs:575]: Yes!

**Comment by BohuTANG** [comment]
Is this PR still working?

**Review by sundy-li — APPROVED** [review]

**Pull request merged**